stdlib/public/core/StringStorage.swift

```diff
@@ -12,308 +12,474 @@
     }
   }
 
-  @inlinable
   @nonobjc
-  var unmanagedView: _UnmanagedString<CodeUnit> {
-    return _UnmanagedString(start: self.start, count: self.count)
-  }
-}
-
-extension _SwiftStringStorage {
-  // Append operations
+  private var isASCII: Bool { return _countAndFlags.isASCII }
 
   @nonobjc
-  internal final func _appendInPlace<OtherCodeUnit>(
-    _ other: _UnmanagedString<OtherCodeUnit>
-  )
-  where OtherCodeUnit : FixedWidthInteger & UnsignedInteger {
-    let otherCount = Int(other.count)
-    _sanityCheck(self.count + otherCount <= self.capacity)
-    other._copy(into: self.unusedBuffer)
-    self.count += otherCount
+  // @opaque
+  internal var _breadcrumbsAddress: UnsafeMutablePointer<_StringBreadcrumbs?> {
+    let raw = Builtin.getTailAddr_Word(
+      start._rawValue,
+      realCapacity._builtinWordValue,
+      UInt8.self,
+      Optional<_StringBreadcrumbs>.self)
+    return UnsafeMutablePointer(raw)
   }
 
+  // The total capacity available for code units. Note that this excludes the
+  // required nul-terminator
   @nonobjc
-  internal final func _appendInPlace(_ other: _UnmanagedOpaqueString) {
-    let otherCount = Int(other.count)
-    _sanityCheck(self.count + otherCount <= self.capacity)
-    other._copy(into: self.unusedBuffer)
-    self.count += otherCount
+  internal var capacity: Int {
+    return realCapacity &- 1
   }
 
+  // The total capacity available for code units. Note that this excludes the
+  // required nul-terminator
   @nonobjc
-  internal final func _appendInPlace<C: Collection>(contentsOf other: C)
-  where C.Element == CodeUnit {
-    let otherCount = Int(other.count)
-    _sanityCheck(self.count + otherCount <= self.capacity)
-    var (remainder, writtenUpTo) =
-      other._copyContents(initializing: self.unusedBuffer)
-    _precondition(remainder.next() == nil, "Collection underreported its count")
-    _precondition(writtenUpTo == otherCount, "Collection misreported its count")
-    count += otherCount
+  private var realCapacity: Int {
+    return Int(bitPattern:
+      _realCapacityAndFlags & _StringObject.Nibbles.largeAddressMask)
   }
 
-  @_specialize(where C == Character._SmallUTF16, CodeUnit == UInt8)
+  // The unused capacity available for appending. Note that this excludes the
+  // required nul-terminator.
+  //
+  // NOTE: Callers who wish to mutate this storage should enfore nul-termination
   @nonobjc
-  internal final func _appendInPlaceUTF16<C: Collection>(contentsOf other: C)
-  where C.Element == UInt16 {
-    let otherCount = Int(other.count)
-    _sanityCheck(self.count + otherCount <= self.capacity)
-    // TODO: Use _copyContents(initializing:) for UTF16->UTF16 case
-    var it = other.makeIterator()
-    for p in end ..< end + otherCount {
-      p.pointee = CodeUnit(it.next()!)
+  private var unusedStorage: UnsafeMutableBufferPointer<UInt8> {
+    @inline(__always) get {
+      return UnsafeMutableBufferPointer(
+        start: mutableEnd, count: capacity)
```

Comment:
I think this should use `unusedCapacity` as the count.

Reply:
Doh, good catch.